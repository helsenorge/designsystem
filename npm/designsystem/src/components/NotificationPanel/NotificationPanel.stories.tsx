import React from 'react';

import { StoryObj, Meta } from '@storybook/react';

import NotificationPanel from './NotificationPanel';
import { allTitleTags } from '../../../.storybook/knobs';
import Docs from '../../docs';
import { getColor } from '../../theme/currys';

const meta = {
  title: '@helsenorge/designsystem-react/Components/NotificationPanel',
  component: NotificationPanel,
  parameters: {
    docs: {
      page: (): React.JSX.Element => <Docs component={NotificationPanel} />,
      description: {
        component: 'Et NotificationPanel lar deg vise viktig informasjon avskilt fra bakgrunnen med ulike farger og ikoner',
      },
    },
  },
  args: {
    dismissable: false,
    size: 'large',
    fluid: false,
    label: 'Det har skjedd noe galt. Prøv igjen senere.',
    labelHtmlMarkup: 'h1',
    variant: 'error',
    role: undefined,
  },
  argTypes: {
    dismissable: {
      control: 'boolean',
    },
    size: {
      control: 'select',
      options: ['small', 'medium', 'large'],
    },
    fluid: {
      control: 'boolean',
    },
    label: {
      control: 'text',
    },
    labelHtmlMarkup: {
      control: 'select',
      options: allTitleTags,
    },
    variant: {
      control: 'select',
      options: ['info', 'warn', 'error', 'success'],
    },
    role: {
      control: 'select',
      options: ['', 'alert', 'region'],
    },
  },
} satisfies Meta<typeof NotificationPanel>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  render: args => <NotificationPanel {...args} label="Label only"></NotificationPanel>,
};

export const WithoutLabel: Story = {
  render: args => (
    <>
      <NotificationPanel {...args}>
        <span>{'Eksempel med label'}</span>
      </NotificationPanel>
      <br />
      <NotificationPanel {...args} label={undefined}>
        <span>{'Eksempel uten label. Nå får innhold en font-vekt som ligner mer på label'}</span>
      </NotificationPanel>
      <br />
      <NotificationPanel
        {...args}
        label={undefined}
        expanderChildren={'Dette er innhold i expanderen. Det vises først når man trykker på knappen.'}
        expanderButtonText="Vis mindre"
        expanderButtonClosedText="Vis mer"
      >
        <span>{'Eksempel uten label. Nå får innhold en font-vekt som ligner mer på label'}</span>
      </NotificationPanel>
    </>
  ),
};

export const AllVariants: Story = {
  render: args => (
    <>
      <div className="row">
        <div className={'col-12'}>
          <NotificationPanel {...args} variant={'info'}>
            <div>
              {`På grunn av kommunesammenslåingen ved nyttår vil enkelte Helsenorge-tjenester være ustabile eller tidsvis utilgjenglig i en
              periode i begynnelsen av januar. Vi beklager ulempnene dette medfører og oppfordrer til å prøve igjen senere.`}
              <a href={'https://www.helsenorge.no'} target="_parent">
                {'Les mer om dine rettigheter her.'}
              </a>
            </div>
          </NotificationPanel>
        </div>
      </div>
      <div className="row mt-6">
        <div className={'col-12'}>
          <NotificationPanel {...args} variant={'warn'}>
            <div>
              {`På grunn av kommunesammenslåingen ved nyttår vil enkelte Helsenorge-tjenester være ustabile eller tidsvis utilgjenglig i en
              periode i begynnelsen av januar. Vi beklager ulempnene dette medfører og oppfordrer til å prøve igjen senere. `}
              <a href={'https://www.helsenorge.no'} target="_blank" rel="noopener noreferrer">
                {'Les mer om dine rettigheter her (external).'}
              </a>
            </div>
          </NotificationPanel>
        </div>
      </div>
      <div className="row mt-6">
        <div className={'col-12'}>
          <NotificationPanel {...args} variant={'error'}>
            <div>
              {`På grunn av kommunesammenslåingen ved nyttår vil enkelte Helsenorge-tjenester være ustabile eller tidsvis utilgjenglig i en
              periode i begynnelsen av januar. Vi beklager ulempnene dette medfører og oppfordrer til å prøve igjen senere. `}
              <a href={'https://www.helsenorge.no'} target="_parent">
                {'Les mer om dine rettigheter her.'}
              </a>
            </div>
          </NotificationPanel>
        </div>
      </div>
      <div className="row mt-6">
        <div className={'col-12'}>
          <NotificationPanel {...args} variant={'success'}>
            <div>
              {`På grunn av kommunesammenslåingen ved nyttår vil enkelte Helsenorge-tjenester være ustabile eller tidsvis utilgjenglig i en
              periode i begynnelsen av januar. Vi beklager ulempnene dette medfører og oppfordrer til å prøve igjen senere. `}
              <a href={'https://www.helsenorge.no'} target="_parent">
                {'Les mer om dine rettigheter her.'}
              </a>
            </div>
          </NotificationPanel>
        </div>
      </div>
    </>
  ),
};

export const Expander: Story = {
  render: args => (
    <>
      <div className="row mt-6">
        <div className={'col-12'}>
          <NotificationPanel
            {...args}
            label={'Med expander'}
            expanderChildren={'Dette er innhold i expanderen. Det vises først når man trykker på knappen.'}
            expanderButtonText="Vis mindre"
            expanderButtonClosedText="Vis mer"
          >
            {'Dette er innhold'}
          </NotificationPanel>
        </div>
      </div>
      <div className="row mt-6">
        <div className={'col-12'}>
          <NotificationPanel
            {...args}
            label={'Med expander'}
            expanderChildren={'Dette er innhold i expanderen. Dette innholdet er åpnet fra start'}
            expanderButtonText="Vis mindre"
            expanderButtonClosedText="Vis mer"
            expanderOpenFromStart
          >
            {'Dette er innhold'}
          </NotificationPanel>
        </div>
      </div>
    </>
  ),
};
export const Dismissable: Story = {
  render: args => (
    <NotificationPanel {...args} label={'Dismissable'} dismissable>
      {'Dette er dismissesable'}
    </NotificationPanel>
  ),
};

export const Compact: Story = {
  render: args => (
    <div style={{ backgroundColor: getColor('blueberry', 50), padding: '3rem' }}>
      <div className="row mt-6">
        <div className={'col-12'}>
          <NotificationPanel {...args} label={'Compact - basic'} compactVariant={'basic'}></NotificationPanel>{' '}
        </div>
      </div>
      <div className="row mt-6">
        <div className={'col-12'}>
          <NotificationPanel {...args} label={'Compact - outline'} compactVariant={'outline'}></NotificationPanel>
        </div>
      </div>
    </div>
  ),
};

export const CompactWithChildren: Story = {
  render: args => (
    <div style={{ backgroundColor: getColor('blueberry', 50), padding: '3rem' }}>
      <div className="row mt-6">
        <div className={'col-12'}>
          <NotificationPanel {...args} label={undefined} compactVariant={'basic'}>
            <div>
              {`På grunn av kommunesammenslåingen ved nyttår vil enkelte Helsenorge-tjenester være ustabile eller tidsvis utilgjenglig i en
              periode i begynnelsen av januar. Vi beklager ulempnene dette medfører og oppfordrer til å prøve igjen senere.`}
              <a href={'https://www.helsenorge.no'} target="_parent">
                {'Les mer om dine rettigheter her.'}
              </a>
            </div>
          </NotificationPanel>{' '}
        </div>
      </div>
      <div className="row mt-6">
        <div className={'col-12'}>
          <NotificationPanel {...args} label={undefined} compactVariant={'outline'}>
            <div>
              {`På grunn av kommunesammenslåingen ved nyttår vil enkelte Helsenorge-tjenester være ustabile eller tidsvis utilgjenglig i en
              periode i begynnelsen av januar. Vi beklager ulempnene dette medfører og oppfordrer til å prøve igjen senere.`}
              <a href={'https://www.helsenorge.no'} target="_parent">
                {'Les mer om dine rettigheter her.'}
              </a>
            </div>
          </NotificationPanel>
        </div>
      </div>
    </div>
  ),
};

export const WithSetWidth: Story = {
  render: args => (
    <>
      <div className="row mt-6">
        <div className={'col-12'}>
          <NotificationPanel {...args} size="small" label={'Size: small'}></NotificationPanel>{' '}
        </div>
      </div>
      <div className="row mt-6">
        <div className={'col-12'}>
          <NotificationPanel {...args} size="medium" label={'Size: medium'}></NotificationPanel>
        </div>
      </div>
      <div className="row mt-6">
        <div className={'col-12'}>
          <NotificationPanel {...args} size="large" label={'Size: large'}></NotificationPanel>
        </div>
      </div>
    </>
  ),
};
