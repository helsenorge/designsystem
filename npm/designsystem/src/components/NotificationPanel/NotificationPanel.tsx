import React from 'react';

import classNames from 'classnames';

import { DetailButton } from './DetailButton/DetailButton';
import { AnalyticsId, IconSize } from '../../constants';
import { useUuid } from '../../hooks/useUuid';
import { getColor } from '../../theme/currys';
import { palette } from '../../theme/palette';
import { getAriaLabelAttributes } from '../../utils/accessibility';
import Close from '../Close';
import Icon from '../Icon';
import CheckFill from '../Icons/CheckFill';
import ErrorSignFill from '../Icons/ErrorSignFill';
import InfoSignFill from '../Icons/InfoSignFill';
import TriangleX from '../Icons/TriangleX';

import styles from './styles.module.scss';

export type NotificationPanelVariants = 'info' | 'warn' | 'alert' | 'success';
export type NotificationCompactVariants = 'basic' | 'outline';
export type NotificationPanelSizes = 'small' | 'medium' | 'large';
export type LabelTags = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'span';

export interface NotificationPanelProps {
  /** Adds custom classes to the element. */
  className?: string;
  /** Adds inner child elements. */
  children?: React.ReactNode;
  /** Adds inner expander elements. */
  expanderChildren?: React.ReactNode;
  /** Text for expanderButton. */
  expanderButtonText?: string;
  /** Text for expanderButton when closed. */
  expanderButtonClosedText?: string;
  /** Makes expander be open from start. */
  expanderOpenFromStart?: boolean;
  /** Changes the visual representation of the notification panel. */
  variant?: NotificationPanelVariants;
  /** Makes the panel more compact. Available in basic and outline. */
  compactVariant?: NotificationCompactVariants;
  /** Sets a fixed size for the content container. */
  size?: NotificationPanelSizes;
  /** Used in combination with dismissiable property to close the notification panel. */
  onClick?: (e?: React.MouseEvent<HTMLElement, MouseEvent>) => void;
  /** Toggles the close button in the top right corner. */
  dismissable?: boolean;
  /** Enables a fluid outer container that spans the entire width of parent. */
  fluid?: boolean;
  /** Sets a label for the notification panel. */
  label?: string;
  /** Changes the underlying element of the label. */
  labelHtmlMarkup?: LabelTags;
  /** Close button aria-label */
  ariaLabelCloseBtn?: string;
  /** Custom id for the label */
  labelId?: string;
  /** Custom role for the panel. Default is no role. If variant is alert or crisis, the aria role will be set to "alert" unless the role-prop is also set. */
  role?: 'region' | 'alert';
  /** Sets the data-testid attribute. */
  testId?: string;
}

type WrapFluidProps = Pick<NotificationPanelProps, 'fluid'> & {
  children: React.ReactElement;
};

const FluidWrapper: React.FC<WrapFluidProps> = ({ fluid, children }) => {
  if (fluid) {
    const fluidClasses = classNames(styles['fluid-wrapper']);

    return <div className={fluidClasses}>{children}</div>;
  }
  return children;
};

const NotificationPanel = React.forwardRef<HTMLDivElement, NotificationPanelProps>((props, ref) => {
  const {
    children,
    variant = 'info',
    dismissable = false,
    onClick,
    expanderChildren,
    expanderButtonText,
    expanderButtonClosedText,
    expanderOpenFromStart = false,
    compactVariant,
    label,
    labelHtmlMarkup = 'h1',
    fluid = false,
    size,
    className,
    labelId,
    role,
    testId,
  } = props;
  const uuid = useUuid(labelId);
  const variantToIconMap = {
    info: (
      <Icon
        svgIcon={InfoSignFill}
        color={palette.blueberry700}
        hoverColor={palette.blueberry700}
        size={compactVariant ? IconSize.XSmall : IconSize.Small}
      />
    ),
    warn: (
      <Icon
        svgIcon={ErrorSignFill}
        color={palette.banana700}
        hoverColor={palette.banana700}
        size={compactVariant ? IconSize.XSmall : IconSize.Small}
      />
    ),
    alert: (
      <Icon
        svgIcon={TriangleX}
        color={palette.cherry700}
        hoverColor={palette.cherry700}
        size={compactVariant ? IconSize.XSmall : IconSize.Small}
      />
    ),
    success: (
      <Icon
        svgIcon={CheckFill}
        color={palette.kiwi900}
        hoverColor={palette.kiwi900}
        size={compactVariant ? IconSize.XSmall : IconSize.Small}
      />
    ),
  };
  const renderContent = (): JSX.Element => {
    const outlineVariant = compactVariant === 'outline';
    const contentClasses = classNames(styles['notification-panel__content']);
    const labelClasses = classNames(styles['notification-panel__label'], {
      [styles['notification-panel__label--no-content']]: !children && !expanderChildren,
      [styles['notification-panel__label--compact']]: !!compactVariant,
      [styles['notification-panel__label--outline']]: outlineVariant,
    });
    const childrenClasses = classNames(styles['notification-panel__children'], {
      [styles['notification-panel__children--with-label']]: label,
      [styles['notification-panel__children--expander-no-label']]: expanderChildren && !label,
      [styles['notification-panel__children--outline']]: outlineVariant,
    });
    const CustomTag = labelHtmlMarkup;

    return (
      <div className={contentClasses} id={!label ? uuid : undefined}>
        {label && (
          <CustomTag className={labelClasses} id={uuid}>
            {label}
          </CustomTag>
        )}
        {children && <div className={childrenClasses}>{children}</div>}
        {expanderChildren && expanderButtonText && expanderButtonClosedText && !compactVariant && (
          <DetailButton
            expanderOpenFromStart={expanderOpenFromStart}
            content={expanderChildren}
            buttonText={expanderButtonText}
            buttonClosedText={expanderButtonClosedText}
          />
        )}
      </div>
    );
  };

  const notificationPanelClasses = classNames(
    styles['notification-panel'],
    styles[`notification-panel--${variant}`],
    size && styles[`notification-panel--${size}`],
    {
      [styles['notification-panel--compact']]: !!compactVariant,
      [styles['notification-panel--compact--basic']]: compactVariant === 'basic',
      [styles['notification-panel--compact--outline']]: compactVariant === 'outline',
      [styles['notification-panel--with-content']]: expanderChildren || (label && children),
      [styles['notification-panel--dismissable']]: dismissable,
    },
    className
  );

  const ariaRole = role || (variant === 'alert' && 'alert') || undefined;
  const ariaLabelAttributes = ariaRole ? getAriaLabelAttributes({ label, id: uuid }) : undefined;

  return (
    <FluidWrapper fluid={fluid}>
      <div
        ref={ref}
        role={ariaRole}
        data-testid={testId}
        data-analyticsid={AnalyticsId.NotificationPanel}
        className={notificationPanelClasses}
        {...ariaLabelAttributes}
      >
        <span className={styles['notification-panel__icon']}>{variantToIconMap[variant]}</span>
        {dismissable && (
          <span className={styles['notification-panel__close']}>
            <Close ariaLabel={props.ariaLabelCloseBtn} onClick={onClick} color={getColor('black')} />
          </span>
        )}
        {renderContent()}
      </div>
    </FluidWrapper>
  );
});

export default NotificationPanel;
