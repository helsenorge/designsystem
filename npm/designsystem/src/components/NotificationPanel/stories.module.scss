@import '../../scss/_breakpoints.scss';
.container {
  margin-left: auto;
  margin-right: auto;
  padding-left: 0.5rem;
  padding-right: 0.5rem;
  width: 100%;
  display: flex;

  @media (min-width: map-get($grid-breakpoints, md)) {
    max-width: 66rem;
  }

  @media (min-width: map-get($grid-breakpoints, xl)) {
    max-width: 84rem;
  }
}
