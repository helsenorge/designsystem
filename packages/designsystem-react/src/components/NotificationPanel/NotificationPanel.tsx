import React from 'react';
import {palette} from '../../theme/palette';
import Icon, {SvgIcon} from '../Icons';
import {useHover} from '../..';
import {
  StyledNotificationPanelFluidWrapper,
  StyledNotificationPanel,
  StyledNotificationPanelIconColumn,
  StyledNotificationPanelContent,
  StyledNotificationPanelActionColumn,
  StyledLabel,
  StyledCloseButton,
} from './NotificationPanel.styled';
import InfoSignStroke from '../Icons/InfoSignStroke';
import AlertSignStroke from '../Icons/AlertSignStroke';
import AlertSignFill from '../Icons/AlertSignFill';
import X from '../Icons/X';

export type NotificationPanelVariants = 'info' | 'warn' | 'alert' | 'crisis';
export type NotificationPanelSizes = 'small' | 'medium' | 'large';

interface NotificationPanelProps {
  className?: string;
  children?: React.ReactNode;
  variant?: NotificationPanelVariants;
  shadow?: boolean;
  size?: NotificationPanelSizes;
  onClick?: (e?: React.MouseEvent<HTMLElement, MouseEvent>) => void;
  dismissable?: boolean;
  fluid?: boolean;
  label?: string;
}

export const variantToColorMap = {
  info: {color: 'kiwi', depth: 50},
  warn: {color: 'banana', depth: 50},
  alert: {color: 'cherry', depth: 100},
  crisis: {color: 'black', depth: 0},
};

export const variantToIconMap = {
  info: <Icon svgIcon={InfoSignStroke} color={palette.kiwi900} hoverColor={palette.kiwi900} />,
  warn: <Icon svgIcon={AlertSignStroke} color="black" hoverColor="black" />,
  alert: <Icon svgIcon={AlertSignFill} color={palette.cherry500} hoverColor={palette.cherry500} />,
  alertLabel: <Icon svgIcon={AlertSignStroke} color={palette.cherry500} hoverColor={palette.cherry500} />,
  crisis: <Icon svgIcon={AlertSignFill} color={palette.banana200} hoverColor={palette.banana200} />,
};

const NotificationPanel = React.forwardRef((props: NotificationPanelProps, ref: any) => {
  const {
    children,
    variant = 'info',
    shadow = false,
    dismissable = false,
    onClick,
    label,
    fluid = false,
    size,
    className,
  } = props;
  const {hoverRef, isHovered} = useHover<HTMLButtonElement>(undefined, dismissable);

  function wrapFluid(panel: React.ReactElement): React.ReactElement {
    if (fluid) {
      return (
        <StyledNotificationPanelFluidWrapper shadow={shadow} variant={variant}>
          {panel}
        </StyledNotificationPanelFluidWrapper>
      );
    }
    return panel;
  }

  return wrapFluid(
    <StyledNotificationPanel
      className={className}
      size={size}
      label={!!label && !children ? true : undefined}
      shadow={!fluid && shadow}
      variant={variant}>
      <StyledNotificationPanelIconColumn>
        {variantToIconMap[variant === 'alert' && label && !children ? 'alertLabel' : variant]}
      </StyledNotificationPanelIconColumn>
      <StyledNotificationPanelContent>
        {label ? <StyledLabel>{label}</StyledLabel> : null}
        {children}
      </StyledNotificationPanelContent>
      <StyledNotificationPanelActionColumn>
        {dismissable ? (
          <StyledCloseButton onClick={onClick} ref={hoverRef}>
            <Icon
              svgIcon={X}
              isHovered={isHovered}
              size={38}
              color={palette.blueberry500}
              hoverColor={palette.blueberry700}
            />
          </StyledCloseButton>
        ) : null}
      </StyledNotificationPanelActionColumn>
    </StyledNotificationPanel>,
  );
});

export default NotificationPanel;
