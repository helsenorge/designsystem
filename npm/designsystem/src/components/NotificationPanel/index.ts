import NotificationPanel from './NotificationPanel';
export * from './NotificationPanel';
export default NotificationPanel;
