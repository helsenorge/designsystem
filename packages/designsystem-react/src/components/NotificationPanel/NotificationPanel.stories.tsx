import React from 'react';
import {withA11y} from '@storybook/addon-a11y';
import {action} from '@storybook/addon-actions';
import {withKnobs, text, select, boolean} from '@storybook/addon-knobs';
import {storiesOf} from '@storybook/react';
import NotificationPanel from './NotificationPanel';

const stories = storiesOf('NotificationPanel', module);
stories.addDecorator(withKnobs);
stories.addDecorator(withA11y);

stories.add('Default', () => (
  <div style={{width: boolean('Fluid', false) ? '100vw' : '1400px'}}>
    <NotificationPanel
      shadow={boolean('Shadow', false)}
      dismissable={boolean('Dismissable', false)}
      onClick={action('on-click-dismiss')}
      size={select('Size', ['small', 'medium', 'large'], 'large')}
      fluid={boolean('Fluid', false)}
      label={text('Label', '')}
      variant={select('Variant', ['info', 'warn', 'alert', 'crisis'], 'alert')}>
      <p>Hallo</p>
    </NotificationPanel>
  </div>
));

stories.add('Simplified (label only)', () => (
  <div style={{width: boolean('Fluid', false) ? '100vw' : '80rem'}}>
    <NotificationPanel
      shadow={boolean('Shadow', false)}
      dismissable={boolean('Dismissable', false)}
      onClick={action('on-click-dismiss')}
      fluid={boolean('Fluid', false)}
      label={text('Label', 'Det har skjedd noe galt. Prøv igjen senere.')}
      variant={select('Variant', ['info', 'warn', 'alert', 'crisis'], 'alert')}
    />
  </div>
));

stories.add('All variants', () => (
  <div className="container">
    <div className="row">
      <div className={select('Columns', ['col-12', 'col-10', 'col-8'], 'col-12')}>
        <NotificationPanel
          shadow={boolean('Shadow', false)}
          dismissable={boolean('Dismissable', false)}
          onClick={action('on-click-dismiss')}
          variant="info"
          label={text('Label', '')}>
          <p>
            På grunn av kommunesammenslåingen ved nyttår vil enkelte Helsenorge-tjenester være ustabile eller tidsvis
            utilgjenglig i en periode i begynnelsen av januar. Vi beklager ulempnene dette medfører og oppfordrer til å
            prøve igjen senere. <a href="/">Les mer om dine rettigheter her.</a>
          </p>
        </NotificationPanel>
      </div>
    </div>
    <div className="row mt-6">
      <div className={select('Columns', ['col-12', 'col-10', 'col-8'], 'col-12')}>
        <NotificationPanel
          shadow={boolean('Shadow', false)}
          dismissable={boolean('Dismissable', false)}
          onClick={action('on-click-dismiss')}
          variant="warn"
          label={text('Label', '')}>
          <p>
            På grunn av kommunesammenslåingen ved nyttår vil enkelte Helsenorge-tjenester være ustabile eller tidsvis
            utilgjenglig i en periode i begynnelsen av januar. Vi beklager ulempnene dette medfører og oppfordrer til å
            prøve igjen senere.{' '}
            <a href="/" target="_blank">
              Les mer om dine rettigheter her (external).
            </a>
          </p>
        </NotificationPanel>
      </div>
    </div>
    <div className="row mt-6">
      <div className={select('Columns', ['col-12', 'col-10', 'col-8'], 'col-12')}>
        <NotificationPanel
          shadow={boolean('Shadow', false)}
          dismissable={boolean('Dismissable', false)}
          onClick={action('on-click-dismiss')}
          variant="alert"
          label={text('Label', '')}>
          <p>
            På grunn av kommunesammenslåingen ved nyttår vil enkelte Helsenorge-tjenester være ustabile eller tidsvis
            utilgjenglig i en periode i begynnelsen av januar. Vi beklager ulempnene dette medfører og oppfordrer til å
            prøve igjen senere. <a href="/">Les mer om dine rettigheter her.</a>
          </p>
        </NotificationPanel>
      </div>
    </div>
    <div className="row mt-6">
      <div className={select('Columns', ['col-12', 'col-10', 'col-8'], 'col-12')}>
        <NotificationPanel
          shadow={boolean('Shadow', false)}
          dismissable={boolean('Dismissable', false)}
          onClick={action('on-click-dismiss')}
          variant="crisis"
          label={text('Label', '')}>
          <p>
            På grunn av kommunesammenslåingen ved nyttår vil enkelte Helsenorge-tjenester være ustabile eller tidsvis
            utilgjenglig i en periode i begynnelsen av januar. Vi beklager ulempnene dette medfører og oppfordrer til å
            prøve igjen senere.{' '}
            <a href="/" target="_blank">
              Les mer om dine rettigheter her.
            </a>
          </p>
        </NotificationPanel>
      </div>
    </div>
  </div>
));
