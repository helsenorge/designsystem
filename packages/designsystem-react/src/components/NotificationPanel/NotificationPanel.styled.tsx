import styled, {css} from 'styled-components';
import {NotificationPanelVariants, variantToColorMap, NotificationPanelSizes} from './NotificationPanel';
import {getColor, getHoverColor} from '../../theme/currys';
import {PaletteNames, PaletteDepths} from '../../theme/palette';
import {typography} from '../../theme/typography';
import {breakpoints} from '../../theme/grid';

interface StyledNotificationPanelProps {
  variant: NotificationPanelVariants;
  size?: NotificationPanelSizes;
  label?: boolean;
  shadow: boolean;
}

const sizeToWidthMap = {
  small: `${breakpoints.sm}px`,
  medium: `${breakpoints.md}px`,
  large: `${breakpoints.lg}px`,
};

const StyledNotificationPanel = styled('div')<StyledNotificationPanelProps>`
  display: grid;
  padding: 1rem;
  grid-template-areas: 'icon' 'content' 'action';
  grid-template-columns: 48px auto 48px;
  grid-template-rows: 1fr;
  max-width: ${props => props.size && sizeToWidthMap[props.size]};
  width: inherit;
  background-color: ${props =>
    getColor(
      variantToColorMap[props.variant].color as PaletteNames,
      variantToColorMap[props.variant].depth as PaletteDepths,
    )};
  ${props =>
    props.variant === 'crisis' &&
    css`
      ${StyledNotificationPanelContent} {
        color: ${getColor('banana', 200)};
        a {
          color: white;
          :hover {
            background-color: rgba(255, 255, 255, 0.2);
            color: white;
          }
          :focus {
            background-color: white;
            color: black;
          }
          &[target='_blank'] {
            /* white icon */
            &:after {
              content: url("data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' width='38px' height='38px' viewBox='0 0 48 48' fill='%23bdbab9'><polygon points='33.305,14.659,33.305,29.205,32.005,29.205,32.005,16.948,15.804,33.149,14.885,32.229,31.155,15.959,18.761,15.96,18.761,14.659' /></svg>");
            }
          }
        }
      }
    `}
  ${props =>
    props.shadow &&
    css`
      box-shadow: 0 0.3rem 0.6rem -0.1rem rgba(0, 0, 0, 0.16);
    `}
    ${props =>
    props.label &&
    css`
      padding: 2rem 1rem;
      ${StyledNotificationPanelContent} {
        ${props.variant === 'alert' &&
        css`
          color: ${getColor('cherry', 500)};
        `};
        display: flex;
        align-items: center;
        margin: 0 1rem;
      }
    `}
`;

const StyledCloseButton = styled('button')`
  border: 0.125rem solid ${getColor('blueberry', 600)};
  background: transparent;
  padding: 0;
  display: flex;
  justify-content: center;
  position: relative;
  outline: none;
  cursor: pointer;
  &:before {
    content: ' ';
    position: absolute;
    top: -0.25rem;
    left: -0.25rem;
    right: -0.25rem;
    bottom: -0.25rem;
    border: 0.125rem solid transparent;
  }
  &:after {
    content: ' ';
    position: absolute;
    top: -0.375rem;
    left: -0.375rem;
    right: -0.375rem;
    bottom: -0.375rem;
    border: 0.125rem solid transparent;
  }
  &:hover,
  :active,
  :focus {
    background-color: ${getHoverColor('blueberry')};
    border-color: transparent;
    &:before {
      border: 0.125rem solid ${getHoverColor('blueberry')};
    }
    &:after {
      border: 0.125rem solid ${getColor('blueberry', 700)};
    }
  }
`;

const StyledNotificationPanelIconColumn = styled('span')`
  grid-area: 'icon';
`;

const StyledNotificationPanelContent = styled('div')`
  grid-area: 'content';
  margin: 0.5rem 1.5rem;
  & p {
    margin: 0;
  }
  & a {
    outline: none;
    padding-bottom: 0.15rem;
    color: ${getColor('blueberry', 600)};
    border-bottom: 1px solid ${getColor('neutral', 300)};
    text-decoration: none;
    &:focus {
      outline: none;
      background-color: rgba(8, 102, 124, 0.05); /* Dette er blueberry 700 med 5% opacity*/
      border-color: ${getColor('neutral', 300)};
    }
    &:hover {
      cursor: pointer;
      border-color: ${getColor('blueberry', 700)};
    }
    &[target='_blank'] {
      &:after {
        content: url("data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' width='38px' height='38px' viewBox='0 0 48 48' fill='%2308667c'><polygon points='33.305,14.659,33.305,29.205,32.005,29.205,32.005,16.948,15.804,33.149,14.885,32.229,31.155,15.959,18.761,15.96,18.761,14.659' /></svg>");
        display: inline-block;
        width: 38px;
        height: 38px;
        vertical-align: bottom;
        margin-bottom: -0.25rem;
      }
    }
  }
`;

const StyledNotificationPanelActionColumn = styled('span')`
  grid-area: 'action';
`;

const StyledLabel = styled('h1')`
  ${typography.label}
  margin: 0;
`;

const StyledNotificationPanelFluidWrapper = styled('div')<StyledNotificationPanelProps>`
  display: flex;
  justify-content: center;
  width: 100%;
  background-color: ${props =>
    getColor(
      variantToColorMap[props.variant].color as PaletteNames,
      variantToColorMap[props.variant].depth as PaletteDepths,
    )};
  ${props =>
    props.shadow &&
    css`
      box-shadow: 0 0.3rem 0.6rem -0.1rem rgba(0, 0, 0, 0.16);
    `}
`;

export {
  StyledNotificationPanel,
  StyledNotificationPanelActionColumn,
  StyledNotificationPanelContent,
  StyledNotificationPanelFluidWrapper,
  StyledNotificationPanelIconColumn,
  StyledLabel,
  StyledCloseButton,
};
