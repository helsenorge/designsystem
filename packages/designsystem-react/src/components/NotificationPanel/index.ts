import NotificationPanel from './NotificationPanel';
export default NotificationPanel;
