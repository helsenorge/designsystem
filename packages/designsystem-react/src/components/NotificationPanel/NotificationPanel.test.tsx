import React from 'react';
import {render} from '@testing-library/react';
import NotificationPanel from './NotificationPanel';

test('displays a notification panel', (): void => {
  const {container} = render(<NotificationPanel>Some text here for testing.</NotificationPanel>);
  expect(container).toMatchSnapshot();
});
