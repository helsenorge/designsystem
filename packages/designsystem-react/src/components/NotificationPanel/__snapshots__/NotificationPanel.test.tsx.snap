// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays a notification panel 1`] = `
<div>
  <div
    class="sc-AxirZ lgMsec"
  >
    <span
      class="sc-AxhCb giigQO"
    >
      <svg
        aria-hidden="true"
        class="sc-AxjAm jPyZsV hnds-style-icon "
        fill="#078141"
        height="48"
        role="img"
        viewBox="0 0 48 48"
        width="48"
      >
        <path
          d="M25.672 30.501a.682.682 0 01-.322-.272.835.835 0 01-.11-.443v-7.917l-.112-.112-3.651.194v.544c.148.015.314.05.495.104.18.055.316.112.406.172a.994.994 0 01.299.316.88.88 0 01.119.459v6.337a.879.879 0 01-.098.443.623.623 0 01-.335.257c-.09.035-.197.06-.32.074a8.883 8.883 0 01-.38.038v.544h4.709v-.544a2.45 2.45 0 01-.7-.194M24 37.604c-7.5 0-13.604-6.103-13.604-13.604 0-7.501 6.104-13.603 13.604-13.603 7.502 0 13.602 6.102 13.602 13.603 0 7.501-6.1 13.604-13.601 13.604m0-28.909C15.56 8.695 8.694 15.562 8.694 24c0 8.438 6.867 15.305 15.305 15.305 8.438 0 15.304-6.867 15.304-15.305 0-8.438-6.866-15.305-15.304-15.305m-.274 10.632a1.559 1.559 0 100-3.118 1.559 1.559 0 000 3.118"
        />
      </svg>
    </span>
    <div
      class="sc-AxhUy dkmWPL"
    >
      Some text here for testing.
    </div>
    <span
      class="sc-AxgMl dnqiUY"
    />
  </div>
</div>
`;
